### DISCUSSION TITLE
Multiple Records Associated with a Single Record

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#667] [Maintainer discussions]
The design for issue #443  is ready for review. Please follow the [Review Guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines) and make sure to review before the set deadline expires.

### Document
https://wiki.mathesar.org/en/design/specs/multiple-records-spec

### Reviewers
@kgodey @pavish @mathemancer  Your review is required.
Feedback from the community is also welcome. Contributors can find more context about Mathesar's design strategy [here](https://wiki.mathesar.org/design).

### Deadline
The deadline for submitting feedback is set for next week, on on Tuesday 28th.

**Comment by kgodey** [comment]
@ghislaineguerin I like the concept here, but we cannot show multiple records in a table. Tables will only have a single value in each column.

So when they say "allow multiple", we will need to:
- create a "track_artist_map" table
- create a new "Artist with Tracks" view with all the original columns of the artist + a summary column called "tracks" set to the "values" option
- redirect them to the new view

The original "artist" table will not have any columns added to it at all, so I'm not sure it makes sense to have the "Allow Multiple" option attached to a column.

**Reply by ghislaineguerin** [reply]
@kgodey got it. I had created that map table, I thought we could actually visualize the mapped contents in that column. I understand now why this might be complicated for users to get. I will modify the prototype to show the scenario as you described it.

**Comment by ghislaineguerin** [comment]
@kgodey 
I want to confirm the correct behavior as we've discussed yesterday.
We established that a user could go from a single record related to multiple from either a view or a table. If they do so from a table, it will prompt them to add a view, at the same time a mapping table will be created.

#### Adding Summary Columns 
The user will have the option to add any available column to a view. In this case, 'releaseName' was initially added by creating a view from the 'release' table first. Once create the user added the second column 'trackName' from the 'track' table.
If there was no mapping table created, then 'trackName' would have been empty with an option to start adding records. 
If there's a map already it will show the mapped values. 

![image](https://user-images.githubusercontent.com/845767/136076546-81481bf1-bdc3-486e-91ab-9a8a1b5d0109.png)

#### Editing a Summary Record
When the user wants to edit, they can choose to remove any of the records. Which will remove the matching row from the 'release_track' table. Or they can choose to add an additional record. Which should swap the list of records for a form in which the user can define the new pair. 

![image](https://user-images.githubusercontent.com/845767/136076583-a026118c-407b-4066-9f54-9836da4e2330.png)

#### Deleting a Mapping Table
If the mapping table is deleted, the summary column will remain in the view but will go back to being blank.

![image](https://user-images.githubusercontent.com/845767/136076617-d33512b2-d837-4b11-9fa3-7c7c07623638.png)

#### Questions
If the view has multiple columns, let's say there's an 'artistName' column and a 'releaseName' column and both are linked by foreign key.
| releaseName | artistName | trackName |
|---|---|---|
| Aquarius | Aqua |  |
| Aquarium | Aqua |  |

So now I add 'trackName', which could result in two different mapping tables:

| releaseName | trackName |
|---|---|
| Aquarius | Around the World |
| Aquarius | Barbie Girl |

| artistName | trackName |
|---|---|
| Aqua | Around the World |
| Aqua | Barbie Girl |

Do we create both, or does the user only map to one of the fields?

**Reply by kgodey** [reply]
@ghislaineguerin The behavior you described seems accurate.

For your question, I think we would need to ask the user which table they'd want to map to since there are two options. In any case, we need to inform the user that we'll be adding a column or creating a map table and get their confirmation before proceeding.

**Comment by ghislaineguerin** [comment]
@kgodey @pavish 
https://www.loom.com/share/5f8be5ec0c014056be4d0a0da83bd081
Please take a look at the video above. I have some questions that we must address before I can close this.

**Reply by pavish** [reply]
Answering your questions @ghislaineguerin:

While the mapping table is created:
  - We could auto-populate the referenced column in the new mapping table. The other column in the mapping table could be marked nullable in the db, so it's a valid db state. I'll defer to @kgodey and @mathemancer to decide on whether we should allow that.
  - The referenced column from the original table would not be removed, since it will act as the column being referenced from the new table.
    - In your example, we would not be creating a new column called 'trackId', instead we will reference the 'id' column directly.
  - A dedicated view will be essential to view the contents of artist and track tables together.

---

Looking at the video, I think I'd be confused when working with Mathesar for the multiple types.
* From a spreadsheet/airtable/any no-code tool perspective, multiple would mean viewing multiple entries like a multi select option in the same table.
* From a DB perspective, we will have a mapping table, and we could have a similar behaviour to no-code tools with an additional view.

Some suggestions I feel would make me understand it better:
1. We could have a separate option in the column dropdown for multiple, instead of showing '+Add multiple' within 'link to another table'.
   - Or we could show it as a separate option inside the modal that opens up. Showing it along with one->one mapping seems to imply that both are similar, while infact they are not.
2. Along with the help text that tells the user that a mapping table will be created, we could allow editing the name of the new table, and have a checkbox to automatically create a new view that combines both.
3. We need to define the UX for the new view, whether each mapping will be a single record or if the cell will contain multiple records, and how it would work.

**Reply by kgodey** [reply]
@ghislaineguerin I looked through the video. I remember us talking through the scenario that you're asking about in our last call. My recollection is that the solution we decided on was:
- Only allow users to add summary fields with multiple records from views, not tables
- If a user wants to "add multiple" from a table, we should guide them towards creating a view first. The flow you have in the video creates a junction table directly, but we want them to create a view. Only adding new columns to views should create junction tables, there should be no way to create a junction table from a table.
- Mapping tables can be pre-populated or not, there's no restriction from the backend on that. This needs to be answered from the design side based on what's intuitive for the user. We may want to support both options (e.g. provide a single button to pre-populate the mapping table rather than pre-populating by default, or ask the user when creating the junction table whether they want data pre-populated).

This should also address your question about "converting" a foreign key column to a multi-record column. This won't be possible at all since we guide them towards creating a view first and only allow "multiple record" summary columns in views. The original field won't be affected.

**Comment by ghislaineguerin** [comment]
@kgodey @pavish Thanks for your comments
I think, based on your feedback that we should consider the following:
- Linking to multiple records should not be a column action, as there's no need to add a column to the table to link multiple records, this will only confuse users.
- We can design the flow in a way that allows users to add records to the junction table without having to look at its contents.
- We can have a summary column that allows displaying multiple records within a single cell
- We don't need to modify the table in any way (no new columns added)

Please see the attached video for a demonstration:
https://www.loom.com/share/e0b4cc02d86e42d0b967d2994218ecdb

If this makes sense I will finalize the design and update the specs. Thanks!

**Reply by kgodey** [reply]
@ghislaineguerin The general flow looks good to me, the only thing I'm not sure about is the placement of "Link to Multiple". I liked the previous placement because it seemed context sensitive, the user would probably think of it while looking at a column. At the table level, it's not as clear what the context is, and i can see the user getting confused. I'm not saying that we need to go back to having it at the column level, but I think we need to think more about when in the flow the user would want that functionality.

**Comment by kgodey** [comment]
Based on the new action plan from the [2021-10-15 meeting](https://wiki.mathesar.org/meeting-notes/october-2021), we're going to update the requirements for issues in the milestone. I'm marking this discussion done, we'll start new discussions for further updates.
